test/src/test/java/org/corfudb/infrastructure/logreplication/LogEntryWriterTest.java

```diff
@@ -0,0 +1,320 @@
+        verifyNumberOfTx();
+
+        int numOpaqueEntries = 3;
+        int topologyConfigId = 5;
+
+        // Send the same payload again
+        LogReplicationEntryMsg lrEntryMsg = generateLogEntryMsg(1, numOpaqueEntries, Address.NON_ADDRESS,
+            topologyConfigId);
+
+        txnContext = Mockito.mock(TxnContext.class);
+        Mockito.doReturn(txnContext).when(metadataManager).getTxnContext();
+        Map<LogReplicationMetadataType, Long> metadataMap = constructMetadataMgrMap(topologyConfigId,
+            numOpaqueEntries, numOpaqueEntries);
+        Mockito.doReturn(metadataMap).when(metadataManager).queryMetadata(txnContext,
+            LogReplicationMetadataType.TOPOLOGY_CONFIG_ID, LogReplicationMetadataType.LAST_SNAPSHOT_STARTED,
+            LogReplicationMetadataType.LAST_SNAPSHOT_APPLIED, LogReplicationMetadataType.LAST_LOG_ENTRY_PROCESSED,
+            LogReplicationMetadataType.LAST_OPAQUE_ENTRY_PROCESSED);
+
+        // Verify that the redundant update was not applied
+        boolean success = logEntryWriter.apply(lrEntryMsg);
+        Assert.assertFalse(success);
+        Mockito.verify(txnContext, times(0)).commit();
+    }
+
+    /**
+     * This test gives 2 LogEntry messages to the Writer, first with sequence numbers(timestamps) [1-3] and then with
+     * sequence numbers [2-4].  It then verifies that sequence numbers 2 and 3 were skipped during the second
+     * apply phase.
+     */
+    @Test
+    public void testLogEntryApplyWithSequenceNumOverlap() {
+
+        // Create a LogEntry message with numOpaqueEntries number of opaque entries.
+        // Each entry corresponds to a transaction
+        int numOpaqueEntries = 3;
+        int topologyConfigId = 5;
+
+        verifyNumberOfTx();
+
+        // Construct the next batch with sequence numbers [2-4]
+        LogReplicationEntryMsg lrEntryMsg = generateLogEntryMsg(2, 4, numOpaqueEntries, topologyConfigId);
+
+        txnContext = Mockito.mock(TxnContext.class);
+
+        Mockito.doReturn(txnContext).when(metadataManager).getTxnContext();
+        Map<LogReplicationMetadataType, Long> metadataMap = constructMetadataMgrMap(topologyConfigId,
+            numOpaqueEntries, numOpaqueEntries);
+
+        Mockito.doReturn(metadataMap).when(metadataManager).queryMetadata(txnContext,
+            LogReplicationMetadataType.TOPOLOGY_CONFIG_ID, LogReplicationMetadataType.LAST_SNAPSHOT_STARTED,
+            LogReplicationMetadataType.LAST_SNAPSHOT_APPLIED, LogReplicationMetadataType.LAST_LOG_ENTRY_PROCESSED,
+            LogReplicationMetadataType.LAST_OPAQUE_ENTRY_PROCESSED);
+
+        boolean success = logEntryWriter.apply(lrEntryMsg);
+        Assert.assertTrue(success);
+
+        // As sequence numbers [1-3] were already applied, only the transaction corresponding to sequence number 4
+        // will be applied.  So the total number of times commit is invoked should be 1
+        Mockito.verify(txnContext, times(1)).commit();
```

Comment:
Since MetadataManager is mocked in this test, it is not possible to verify the sequence numbers.  I can add the verification in a similar test in MetadataManagerTest.